internal/k8s/client.go

```diff
@@ -314,42 +339,141 @@ func (k K8sClient) Upsert(ctx context.Context, entities []K8sEntity, timeout tim
+
+	err := resources.Visit(func(info *resource.Info, err error) error {
+		if err != nil {
+			return err
+		}
+
+		helper := resource.NewHelper(info.Client, info.Mapping)
+		existing, err := helper.Get(info.Namespace, info.Name)
+		if err != nil {
+			if apierrors.IsNotFound(err) {
+				return nil
+			}
+			return errors.Wrap(err, "could not get information about the resource")
+		}
+
+		newRes := *info
+		_ = newRes.Refresh(existing, true)
+
+		requireUpdate.Append(&newRes)
+		return nil
+	})
+
+	return requireUpdate, err
+}
+
+func (k K8sClient) helmResultToEntities(result *kube.Result) ([]K8sEntity, error) {
+	entities := []K8sEntity{}
+	for _, info := range result.Created {
+		entities = append(entities, NewK8sEntity(info.Object))
+	}
+	for _, info := range result.Updated {
+		entities = append(entities, NewK8sEntity(info.Object))
+	}
+
+	// Helm parses the results as unstructured info, but Tilt needs them parsed with the current
+	// API scheme. The easiest way to do this is to serialize them to yaml and re-parse again.
+	buf, err := SerializeSpecYAMLToBuffer(entities)
+	if err != nil {
+		return nil, errors.Wrap(err, "reading k8s result")
+	}
+
+	parsed, err := ParseYAML(buf)
+	if err != nil {
+		return nil, errors.Wrap(err, "parsing k8s result")
+	}
+	return parsed, nil
 }
 
 // applyEntityAndMaybeForce `kubectl apply`'s the given entity, and if the call fails with
 // an immutible field error, attempts to `replace --force` it.
 func (k K8sClient) applyEntityAndMaybeForce(ctx context.Context, entity K8sEntity) ([]K8sEntity, error) {
-	stdout, stderr, err := k.actOnEntity(ctx, []string{"apply", "-o", "yaml"}, entity)
+	existing, resources, err := k.prepareUpdate(ctx, []K8sEntity{entity})
 	if err != nil {
-		reason, shouldTryReplace := maybeShouldTryReplaceReason(stderr)
+		return nil, errors.Wrap(err, "kubectl apply")
+	}
 
+	result, err := k.helmKubeClient.Update(existing, resources, false)
+	if err != nil {
+		reason, shouldTryReplace := maybeShouldTryReplaceReason(err.Error())
 		if !shouldTryReplace {
-			return nil, errors.Wrapf(err, "kubectl apply:\nstderr: %s", stderr)
+			return nil, err
 		}
 
-		// NOTE(maia): we don't use `kubecutl replace --force`, because we want to ensure that all
+		// NOTE(maia): we don't use `kubectl replace --force`, because we want to ensure that all
 		// dependant pods get deleted rather than orphaned. We WANT these pods to be deleted
 		// and recreated so they have all the new labels, etc. of their controlling k8s entity.
 		logger.Get(ctx).Infof("Applying %s failed. Retrying with 'kubectl delete && create': %s", entity.Name(), reason)
-		// --ignore-not-found because, e.g., if we fell back due to large metadata.annotations, the object might not exist
-		_, stderr, err = k.actOnEntity(ctx, []string{"delete", "--ignore-not-found=true"}, entity)
-		if err != nil {
-			return nil, errors.Wrapf(err, "kubectl delete (as part of delete && create):\nstderr: %s", stderr)
+		if len(existing) > 0 {
```

Comment:
Doesn't hurt, but just to check my understanding - IIUC we can only get here if `existing == 0`. Have I missed a case or is this just defensive programming?

Reply:
mostly defensive -- once we rip out the kubectl runner, i expect this is going to get refactored a lot (since the current code flow doesn't really make sense if there's no kubectl runner)